### ISSUE TITLE
Question about --singletons

### ISSUE THREAD

**Issue by mirimarine** [jtamames/SqueezeMeta#238]
Hi all,
I have a question concerning the new singletons mode in SqueezeMeta.
In the manual it says that using this mode would result in 100% mapping statistics. 
I have tried out this mode twice by now, and it seems to me like mapping is still only performed on the real contigs but not the singletons.
My mapping statistics look like this:
```
Sample	        Total reads	Mapped reads	    Mapping perc	Total bases
metagenome	57977	        37376	            64.47	                184630499
```
When I check the contigtable and orftable in the results folder, I find that all singletons have 0 read counts.
As we would like to include transcriptome data in our SQM run, it would be great if mapping would also include the singleton sequences. 
As indexing for mapping is only done once (before identification of the singletons), I thought that I could stop the SQM after step 9, manually perform indexing on the fasta file that includes both contigs and singletons, replace the former index file by the new one and restart SQM from step 10. Do you think that would be possible or do you have any other suggestion how to get mapping statistics of the singletons?

Thank you very much for your help.
All the best,
Miriam

**Comment by jtamames** [comment]
Hello
It should work mapping all reads into all contigs, including singletons. It is difficult to say where the problem is, but a likely explanation is that the singletons were not copied to the contigs' fasta file. Could you send me the syslog and the 22*stats file for me to take a look?

Take into account that indexing is also done in step 10, so it will overwrite your previous index. You would also need to cancel the indexing in step 10 at least. But let´s try to fix the issue first.

Best,

Javier

**Comment by jtamames** [comment]
Heh,  I just realize that what I said in the previous post is not true.. Step 10 checks if there is an index already in place, and if so, it does not make the indexing again. So it should work. But as I said, let´s find out where the issue is.

Best,

**Comment by mirimarine** [comment]
Hi Javier,
thank you for your fast reply.
Here are the files.
All the best,
Miriam
[22.SampleO_sub.stats.zip](https://github.com/jtamames/SqueezeMeta/files/5918170/22.SampleO_sub.stats.zip)
[syslog.zip](https://github.com/jtamames/SqueezeMeta/files/5918171/syslog.zip)

**Comment by jtamames** [comment]
Hello
Just for understanding your experiment: I see you used an external assembly (smpO.contigs.sub2.fasta). Was this done using coassemblying the reads you are providing (these in "metagenome" and "metatranscriptome" samples)?  The reference assembly is very small (51 contigs), is this a draft genome or similar? Also, the "metagenome" data set is also very small (58.000 reads), are these reads real short reads from sequencing, or rather some kind of contigs or genomic fragments? 

Best,
Javier

**Comment by mirimarine** [comment]
Hi Javier,
This was a test run that I performed with a tiny subset of my actual data to double check how the singleton mode works, as I have observed the same behavior on a previous run.
In brief, I did an assembly with one metagenome. From the assembly, I kept 100 contigs. Then I mapped back my metagenome reads and the metatranscriptome reads to these 100 contigs. (The metatranscriptome comes from the same biological sample, from which we did a co-extraction of DNA and RNA for sequencing). I created read files which only contained the reads mapping to the 100 contigs. Then I discarded 50 contigs, to make sure that I have some "singleton" reads.
I run SQM with the subset metagenome assembly (50 contigs) and the metagenome and metatranscriptome reads (which mapped to 100 contigs).
I hope I explained this in an understandable way. Please let me know if you have any questions.
Cheers,
Miriam

**Comment by jtamames** [comment]
Hello

Yes, very well explained. Apparently the singletons are being extracted correctly, as the stats file shows, but the mapping step in 10 is failing. Could you paste here the full `10*mappingstat` and `intermediate/01*mappingstat_orig` files? Also, could you reproduce the issue using Bowtie instead of bwa?

Best,

**Comment by mirimarine** [comment]
Hi,
Here are the files that you asked for. 
I just started a run with the same data and settings, except bowtie mapping instead of bwa. However, I don't know how bowtie behaves with PacBio long reads.
I will let you know how this worked out once it has finished (probably by tomorrow).
All the best,
Miriam
[01.SampleO_sub.mappingstat_orig.zip](https://github.com/jtamames/SqueezeMeta/files/5919726/01.SampleO_sub.mappingstat_orig.zip)
[10.SampleO_sub.mappingstat.zip](https://github.com/jtamames/SqueezeMeta/files/5919727/10.SampleO_sub.mappingstat.zip)

**Comment by jtamames** [comment]
Ok, so this is PacBio, you said it before but I didn't get it. How did you do the assembly and the mapping? Could it be that you included reads mapping partially to a contig? That would explain things. It is likely that BWA is not doing a good work when aligning long reads.

**Comment by mirimarine** [comment]
Good morning,
Assembly was done with canu, using only the PacBio metagenome reads. Mapping was done with bwa, and I used the same settings as SqueezeMeta does (and the same installation, as I run mapping the SqueezeMeta conda environment).
```
bwa index -p smpO_sub1 smpO.contigs.sub1.fasta
bwa mem smpO_sub1 ../01_rawreads/4656_O_CCS.fastq -v 1 -t 35 > mg.sam
bwa mem smpO_sub1 ../01_rawreads/transcriptomes/O_dep_trimmed_l.fastq.gz -v 1 -t 35 > mt.sam
```
For comparison, I will do mapping of the read files to the contigs+singleton.fasta with bwa externally. Lets see if I get the same result.


The bowtie run finished, but mapping stats look even worse.
```
# Sample        Total reads     Mapped reads    Mapping perc    Total bases
metagenome      48707   		14211   		29.18   		182314469
metatranscriptome 16026170        2039117 		12.72   		2229221924
```
What I forgot to mention earlier, the transcriptome data is single-end, 150bp reads. So all singletons from the transcriptome are filtered out.

All the best,
Miriam

**Comment by mirimarine** [comment]
Hi,
I just did external bwa mapping to fasta containing both contigs and singletons.
In the project directory and with the SqueezeMeta conda environment activated, I ran:
```
mkdir external
bwa index -p external/smpO_cs.index results/01.SampleO_sub.fasta
bwa mem external/smpO_cs.index data/raw_fastq/mgreads.fastq -v 1 -t 35 > external/mg.sam
bwa mem external/smpO_cs.index data/raw_fastq/mtreads.fastq.gz -v 1 -t 35 > external/mt.sam
```
The mapping stats look really good:
metagenome: 99.99% reads mapping
metatranscriptome: 99.48% reads mapping

So I would assume that in the SqueezeMeta run it was not mapping to the fasta file containing both contigs and singletons.

I hope this helps. Please let me know if I should try anything else.

All the best, 
Miriam

**Comment by jtamames** [comment]
Ok, very nice info. Did you check if the sam you got this way is identical to the one of the SqueezeMeta run (in `data/sam` directory)? If it is, then the problem in in the counting. If not, it is in the alignment.
Best,
J

**Comment by mirimarine** [comment]
The mapping files are not identical.
The ones from the external mapping are larger.

I also did the same external mapping, now mapping to the contigs only (so the file that I provided SqueezeMeta as external assembly). 
The mapping stats look like this:
metagenome: 64.47% reads mapping
metatranscriptome: 91.02% reads mapping
This is basically the stats that I got from the SQM run. But the files are still not identical.

**Comment by jtamames** [comment]
Ok, could you do the following in both SAM files, and tell me the result?

grep -c "^@SQ" <sam file>
wc <sam file>

Thanks!

**Comment by mirimarine** [comment]
For the SqueezeMeta internat sam files:
```
grep -c "^@SQ" SampleO_sub.metagenome.sam 
50
wc SampleO_sub.metagenome.sam
58028    843936 384371508 SampleO_sub.metagenome.sam
grep -c "^@SQ" SampleO_sub.tatranscriptome.sam 
50
wc SampleO_sub.tatranscriptome.sam
16078831  238518565 6279897098 SampleO_sub.tatranscriptome.sam
```
For the external mapping to the contigs+singletons:
```
grep -c "^@SQ" mg.sam
20651
wc mg.sam
84081   1050174 394387797 mg.sam
grep -c "^@SQ" mt.sam
20651
wc mt.sam
16707039  261290803 7125088552 mt.sam
```
For the external mapping to the only the contigs:
```
grep -c "^@SQ" mg_contigs.sam
50
wc mg_contigs.sam
58028    843936 384447847 mg_contigs.sam
grep -c "^@SQ" mt_contigs.sam
50
wc mt_contigs.sam
16078831  238518565 6295434799 mt_contigs.sam
```

**Comment by jtamames** [comment]
Ok, then the issue is clear: Somehow the reference is not being done adding the reads but with the initial contigs instead. I will check on this. Thank you very much for your help!

**Comment by jtamames** [comment]
One more thing: You say that the transcripts are 150 bps, then they will not be considered (by default SqueezeMeta drops contigs < 200 bps). Would that be the case also with the metagenomic reads? Could you try to run the test using `-c 0`, which will keep all contigs no matter its size?

**Comment by jtamames** [comment]
Maybe it is better you sending me your project via wetransfer, for not having to disturb you asking for things

**Comment by mirimarine** [comment]
Hi,

**Comment by jtamames** [comment]
Yes please, then you should get mapping stats close to 100%, I hope. I think we must put a warning or something to tell the user that most reads are being dumped and will not be considered as contigs.
Best,
Javier

**Comment by mirimarine** [comment]
Hi  Javier,
here are the results, using `-c 0`.
```
intermediate/01.SampleO_sub_c0.mappingstat_orig
# Sample        Total reads     Mapped reads    Mapping perc
metagenome      48707   28106   57.70
tatranscriptome 16026170        14583044        91.00

# This result corresponds to the mapping of reads to the original assembly, before the introduction of singletons
````

```
results/10.SampleO_sub_c0.mappingstat
# Sample        Total reads     Mapped reads    Mapping perc    Total bases
metagenome      57977   37376   64.47   184630499
tatranscriptome 16078780        14635654        91.02   2231647494
```

Which project would you like to have a look at? The original (bwa mapping), the bowtie or the bwa with -c 0 ?
I will send you an email with a link for downloading.

All the best,
Miriam

**Comment by jtamames** [comment]
The original, please
Best,
Javier

**Comment by jtamames** [comment]
Hello
Ok, I have found the problem. The thing is: For identifying singletons, 01.remap.pl makes an alignment using Bowtie, BWA or MiniMap, as selected by the user. Then the reads not mapping the contigs are introduced into the contig file. So far so good, and you have the correct number of contigs+singletons in the 01.fasta file. The alignment files, including the SAM, are kept. But then 10.mapsamples.pl, when trying to map reads for estimating abundances, first checks if the SAM is already in place. If it is, it assumes that you are re-running that part of the analysis and does not redo the alignment. As a result, you end up having the same result than with no singletons.
When I removed the reference and SAM file and ran 10.mapsamples.pl, I got the correct result:

>  Sample	Total reads	Mapped reads	Mapping perc	Total bases
metagenome	63429	63421	99.99	186610656
tatranscriptome	16078780	14635654	91.02	2231647494

and just 8 reads fail to align because of some alignment issue.

The solution is to add a line at the end of the `01.remap.pl` script, removing the old reference and sam. Put this at the end of the script (line 224):
`system("rm $bowtieref.1.bt2; rm $bowtieref.bwt; rm $outsam");`

Thanks for helping us to spot this one!
Best,
Javier
